crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs

```diff
@@ -174,3 +196,70 @@ pub(crate) fn is_singledispatch_implementation(
+/// - When quoting `kubernetes` in `kubernetes.SecurityContext`, we want `"kubernetes.SecurityContext"`.
+/// - When quoting `Series` in `Series["pd.Timestamp"]`, we want `"Series[pd.Timestamp]"`.
+/// - When quoting `Series` in `Series[Literal["pd.Timestamp"]]`, we want `"Series[Literal['pd.Timestamp']]"`.
+///
+/// In general, when expanding a component of a call chain, we want to quote the entire call chain.
+pub(crate) fn quote_annotation(
+    node_id: NodeId,
+    semantic: &SemanticModel,
+    locator: &Locator,
+    stylist: &Stylist,
+) -> Result<Edit> {
+    let expr = semantic.expression(node_id);
+    if let Some(parent_id) = semantic.parent_expression_id(node_id) {
+        match semantic.expression(parent_id) {
+            Expr::Subscript(parent) => {
+                if expr == parent.value.as_ref() {
+                    // If we're quoting the value of a subscript, we need to quote the entire
+                    // expression. For example, when quoting `DataFrame` in `DataFrame[int]`, we
+                    // should generate `"DataFrame[int]"`.
+                    return quote_annotation(parent_id, semantic, locator, stylist);
+                }
+            }
+            Expr::Attribute(parent) => {
+                if expr == parent.value.as_ref() {
+                    // If we're quoting the value of an attribute, we need to quote the entire
+                    // expression. For example, when quoting `DataFrame` in `pd.DataFrame`, we
+                    // should generate `"pd.DataFrame"`.
+                    return quote_annotation(parent_id, semantic, locator, stylist);
+                }
+            }
+            Expr::Call(parent) => {
+                if expr == parent.func.as_ref() {
+                    // If we're quoting the function of a call, we need to quote the entire
+                    // expression. For example, when quoting `DataFrame` in `DataFrame()`, we
+                    // should generate `"DataFrame()"`.
+                    return quote_annotation(parent_id, semantic, locator, stylist);
+                }
+            }
+            Expr::BinOp(parent) => {
+                if parent.op.is_bit_or() {
+                    // If we're quoting the left or right side of a binary operation, we need to
+                    // quote the entire expression. For example, when quoting `DataFrame` in
+                    // `DataFrame | Series`, we should generate `"DataFrame | Series"`.
+                    return quote_annotation(parent_id, semantic, locator, stylist);
+                }
+            }
+            _ => {}
+        }
+    }
+
+    let annotation = locator.slice(expr);
+
+    // If the annotation already contains a quote, avoid attempting to re-quote it.
+    if annotation.contains('\'') || annotation.contains('"') {
+        return Err(anyhow::anyhow!("Annotation already contains a quote"));
+    }
```

Comment:
In the function example, we've mentioned to update the internal quote:

> - When quoting `Series` in `Series[Literal["pd.Timestamp"]]`, we want `"Series[Literal['pd.Timestamp']]"`.

With this code, I think that doesn't work, right? Can we use the `Stylist` to only check for the opposite quote?

Reply:
Yeah sorry, it doesn't do that yet.